rfc/2022/2022-09-15_Design-System-Team_Migrate-vets-website-off-of-formation.md

```diff
@@ -0,0 +1,106 @@
+
+* Comment Deadline: `2022-10-11`
+* Team Crew and Name: Platform/DST
+* Authors:
+  * [Brooks Johnson](https://github.com/bkjohnson)
+  * [Contributing Author 1 Github Handle](https://github.com/ContributingAuthor1Handle)
+  * [Contributing Author 2 Github Handle](https://github.com/ContributingAuthor2Handle)
+* [Original RFC Pull Request](https://github.com/department-of-veterans-affairs/va.gov-platform-arch/pull/-1)
+
+## Background
+
+`formation`, the CSS framework for va.gov, is dependent on USWDS version 1 as is the rest of va.gov. This is now a dead branch and the latest version of USWDS is version 3.
+In order to continue to receive updates from USWDS, we need to switch to a more recent version.
+
+## Motivation
+
+The Design System team is developing a new CSS library which is based on USWDS3 and may contain additional VA-specific styling.
+By relying on USWDS internals, we are able to provide a lot of utilities while having to maintain less SASS ourselves.
+We also want to namespace all of the classes we provide rather than allow developers full unlimited access to everything within USWDS.
+This gives the Design System more control and will also make future maintenance and updates easier.
+
+We also hope to be able to improve performance with this new css library in a number of ways, including:
+
+- smaller stylesheets due to elimination of `!important` declaration
+- smaller stylesheets due to simpler utility names (`vads-u-background-color--gray` => `vads-bg-gray`)
+
+(Note: these name changes are why teams will need to update their apps to use the names of the new classes)
+
+We'd also like to explore the possibility of allowing teams to rely less on global site-wide assets and declare a minimum subset of which DS assets they will need. Much of this work is out of scope for this RFC.
+
+## Design
+
+Before this migration can begin there are prerequisites which must happen.
+
+### Prerequisites
+
+1. Application, sitewide, & platform code will need to be using proper VADS web components. This means no React components,
+and no CSS class components as these rely on `formation` components.
+    - This will unblock those modules from being removed in formation
+1. We will need a published version of the `css-library` package
+1. Write a new ESLint rule which can trigger when a `vads-u-*` class is used.
+
+Once these have been met we can move on to beginning the migration
+
+### Migration
+
+In order to begin migrating we will need to make the `css-library` available, and this can be done in one of two ways.
+
+1. Add the stylesheets to the sitewide SASS file.
+    - This will cause every page to download the file whether they use the utilities or not
+    - Could be a "final state" of the migration
+1. Add the stylesheets to each specific app's SASS file as they are migrated.
+    - This will bloat the app's bundle size, but other pages and apps won't be affected
```

Comment:
The overall bundle is bigger? I had though that if an app fully switches over to the new css-library that they would not need formation. True or false?

Reply:
That is the end goal, but the plan is to add the new stylesheet with the new names using one of these approaches. Removing formation will only come later once app code has switched to using the new classes and we are sure that formation is no longer needed.

Reply:
Right and with this second approach of adding to a specific app I would expect the app devs to ensure that they were no longer using formation and do the class-change work at the time of adopting the new packages. Otherwise I'm not sure what "migration" means for app teams. I suppose an app team could choose to pull down and run both formation and the new css-library if they wanted to but we should be clear here that it's their decision to do so and we'd not advise they run in that configuration for long as there is a clear performance penalty.